File: StarfallUnity/Assets/Scripts/Core/Defs.cs
using System;
using System.Collections.Generic;

namespace Starfall.Core {

public enum Element { Kinetic = 0, Ember = 1, Surge = 2, Null = 3 }
public enum Rarity { Common = 0, Uncommon = 1, Rare = 2, Legendary = 3, Exotic = 4 }
public enum AmmoType { Primary = 0, Special = 1, Heavy = 2 }
public enum Slot { Kinetic = 0, Energy = 1, Power = 2, Helmet = 3, Arms = 4, Chest = 5, Legs = 6, ClassItem = 7 }
public enum ItemKind { Weapon = 0, Armor = 1 }
public enum StatId { Resilience = 0, Mobility = 1, Recovery = 2, Discipline = 3, Intellect = 4, Strength = 5 }
public enum Rank { Minor = 0, Major = 1, Ultra = 2, Boss = 3 }
public enum RewardTier { World = 0, Powerful = 1, Pinnacle = 2 }

public struct Rgb {
    public float R, G, B;
    public Rgb(float r, float g, float b) { R = r; G = g; B = b; }
}

public sealed class ElementDef {
    public Element Id;
    public string Name;
    public Rgb Color;
    public string Hex;
}

public sealed class RarityDef {
    public Rarity Id;
    public string Name;
    public string Hex;
    public float Weight;     // relative drop likelihood
    public int Perks;        // rolled perk count
    public int StatBonus;    // added to every rolled weapon stat
    public int Shards;       // dismantle base value
}

/// <summary>Static tables shared by the loot economy and the combat maths.</summary>
public static class Defs {
    // ---------------------------------------------------------------- elements
    public static readonly ElementDef[] Elements = {
        new ElementDef { Id = Element.Kinetic, Name = "Kinetic", Hex = "#dce1ee", Color = new Rgb(0.86f, 0.88f, 0.94f) },
        new ElementDef { Id = Element.Ember,   Name = "Ember",   Hex = "#ff7a3c", Color = new Rgb(1.00f, 0.45f, 0.15f) },
        new ElementDef { Id = Element.Surge,   Name = "Surge",   Hex = "#4fd7ff", Color = new Rgb(0.30f, 0.82f, 1.00f) },
        new ElementDef { Id = Element.Null,    Name = "Null",    Hex = "#b47dff", Color = new Rgb(0.66f, 0.44f, 1.00f) },
    };
    public static ElementDef Of(Element e) => Elements[(int)e];
    public static readonly Element[] EnergyElements = { Element.Ember, Element.Surge, Element.Null };

    // ---------------------------------------------------------------- rarity
    public static readonly RarityDef[] Rarities = {
        new RarityDef { Id = Rarity.Common,    Name = "Common",    Hex = "#c9d2e0", Weight = 42f,  Perks = 0, StatBonus = 0,  Shards = 1 },
        new RarityDef { Id = Rarity.Uncommon,  Name = "Uncommon",  Hex = "#6ad07a", Weight = 30f,  Perks = 1, StatBonus = 6,  Shards = 2 },
        new RarityDef { Id = Rarity.Rare,      Name = "Rare",      Hex = "#4f9bff", Weight = 19f,  Perks = 2, StatBonus = 12, Shards = 5 },
        new RarityDef { Id = Rarity.Legendary, Name = "Legendary", Hex = "#b57bff", Weight = 8.2f, Perks = 3, StatBonus = 20, Shards = 14 },
        new RarityDef { Id = Rarity.Exotic,    Name = "Exotic",    Hex = "#f5d33f", Weight = 0.8f, Perks = 0, StatBonus = 26, Shards = 40 },
    };
    public static RarityDef Of(Rarity r) => Rarities[(int)r];

    // ---------------------------------------------------------------- slots
    public static readonly Slot[] WeaponSlots = { Slot.Kinetic, Slot.Energy, Slot.Power };
    public static readonly Slot[] ArmorSlots = { Slot.Helmet, Slot.Arms, Slot.Chest, Slot.Legs, Slot.ClassItem };
    public static readonly Slot[] AllSlots = {
        Slot.Kinetic, Slot.Energy, Slot.Power,
        Slot.Helmet, Slot.Arms, Slot.Chest, Slot.Legs, Slot.ClassItem
    };
    public static bool IsWeaponSlot(Slot s) => s == Slot.Kinetic || s == Slot.Energy || s == Slot.Power;

    public static string SlotName(Slot s) {
        switch (s) {
            case Slot.Kinetic: return "Kinetic";
            case Slot.Energy: return "Energy";
            case Slot.Power: return "Power";
            case Slot.Helmet: return "Helmet";
            case Slot.Arms: return "Gauntlets";
            case Slot.Chest: return "Chest Armor";
            case Slot.Legs: return "Leg Armor";
            default: return "Class Item";
        }
    }

    // ---------------------------------------------------------------- stats
    public static readonly StatId[] Stats = {
        StatId.Resilience, StatId.Mobility, StatId.Recovery,
        StatId.Discipline, StatId.Intellect, StatId.Strength
    };
    public const int StatCount = 6;

    public static string StatName(StatId s) {
        switch (s) {
            case StatId.Resilience: return "Resilience";
            case StatId.Mobility: return "Mobility";
            case StatId.Recovery: return "Recovery";
            case StatId.Discipline: return "Discipline";
            case StatId.Intellect: return "Intellect";
            default: return "Strength";
        }
    }

    public static string StatDescription(StatId s) {
        switch (s) {
            case StatId.Resilience: return "Raises maximum health and reduces flinch.";
            case StatId.Mobility: return "Increases movement and strafe speed.";
            case StatId.Recovery: return "Shields recharge sooner and faster.";
            case StatId.Discipline: return "Reduces grenade cooldown.";
            case StatId.Intellect: return "Reduces Super cooldown.";
            default: return "Reduces melee cooldown.";
        }
    }

    /// <summary>Stat tier 0-10 from a 0-100 total. Tiers are what actually drive the curves.</summary>
    public static int Tier(int statValue) {
        int t = statValue / 10;
        return t < 0 ? 0 : (t > 10 ? 10 : t);
    }

    // ---------------------------------------------------------------- power
    public const int StartPower = 100;
    public const int SoftCap = 300;      // world drops stop climbing here
    public const int PowerfulCap = 350;  // "powerful" rewards climb to here
    public const int PinnacleCap = 370;  // pinnacle rewards only

    /// <summary>
    /// Outgoing damage multiplier from the power delta. Being under the
    /// recommended power bites hard; being over helps, with a ceiling.
    /// </summary>
    public static float DamageOut(int playerPower, int activityPower) {
        int d = playerPower - activityPower;
        float m = d >= 0 ? 1f + d * 0.010f : 1f + d * 0.022f;
        return m < 0.20f ? 0.20f : (m > 1.55f ? 1.55f : m);
    }

    /// <summary>Incoming damage multiplier from the power delta.</summary>
    public static float DamageIn(int playerPower, int activityPower) {
        int d = playerPower - activityPower;
        float m = d >= 0 ? 1f - d * 0.007f : 1f - d * 0.020f;
        return m < 0.55f ? 0.55f : (m > 2.6f ? 2.6f : m);
    }

    /// <summary>Experience needed to advance from `level` to the next.</summary>
    public static int XpForLevel(int level) =>
        (int)(600 + level * 340 + Math.Pow(level, 1.75) * 24);

    /// <summary>Power of a dropped item, given the reward tier.</summary>
    public static int DropPower(int playerPower, RewardTier tier, Rng rng) {
        switch (tier) {
            case RewardTier.Pinnacle:
                return Math.Min(PinnacleCap, playerPower + 3 + rng.RangeInt(0, 2));
            case RewardTier.Powerful:
                return Math.Min(PowerfulCap, playerPower + 2 + rng.RangeInt(0, 2));
            default:
                // World drops are mostly sidegrades, hard-stopped at the soft cap.
                return Math.Min(SoftCap, Math.Max(1, playerPower + rng.RangeInt(-3, 2)));
        }
    }

    // Elemental shields: matching element shreds them, mismatched barely dents them.
    public const float ShieldMatchMultiplier = 3.0f;
    public const float ShieldMismatchMultiplier = 0.45f;
}
}

File: web-prototype/src/data/defs.js
// Shared vocabulary for the loot/combat systems: elements, rarities, ammo,
// character stats and the power-level curves that tie difficulty to gear.

export const ELEMENTS = {
  kinetic: { id: 'kinetic', name: 'Kinetic', color: [0.86, 0.88, 0.94], css: '#dce1ee', glow: [1, 1, 1] },
  ember:   { id: 'ember',   name: 'Ember',   color: [1.0, 0.45, 0.15], css: '#ff7a3c', glow: [1.0, 0.55, 0.18] },
  surge:   { id: 'surge',   name: 'Surge',   color: [0.30, 0.82, 1.0], css: '#4fd7ff', glow: [0.45, 0.9, 1.0] },
  null:    { id: 'null',    name: 'Null',    color: [0.66, 0.44, 1.0], css: '#b47dff', glow: [0.72, 0.5, 1.0] },
};
export const ELEMENT_IDS = ['ember', 'surge', 'null'];

export const RARITY = {
  common:    { id: 'common',    name: 'Common',    css: '#c9d2e0', color: [0.79, 0.82, 0.88], weight: 42, perks: 0, statBonus: 0,  power: -4, shards: 1 },
  uncommon:  { id: 'uncommon',  name: 'Uncommon',  css: '#6ad07a', color: [0.42, 0.82, 0.48], weight: 30, perks: 1, statBonus: 6,  power: -2, shards: 2 },
  rare:      { id: 'rare',      name: 'Rare',      css: '#4f9bff', color: [0.31, 0.61, 1.0],  weight: 19, perks: 2, statBonus: 12, power: 0,  shards: 5 },
  legendary: { id: 'legendary', name: 'Legendary', css: '#b57bff', color: [0.71, 0.48, 1.0],  weight: 8.2, perks: 3, statBonus: 20, power: 1,  shards: 14 },
  exotic:    { id: 'exotic',    name: 'Exotic',    css: '#f5d33f', color: [0.96, 0.83, 0.25], weight: 0.8, perks: 4, statBonus: 26, power: 2,  shards: 40 },
};
export const RARITY_ORDER = ['common', 'uncommon', 'rare', 'legendary', 'exotic'];

export const AMMO = {
  primary: { id: 'primary', name: 'Primary', css: '#e8eefc', color: [0.91, 0.93, 0.99], max: 999, pickup: 0 },
  special: { id: 'special', name: 'Special', css: '#7fe08a', color: [0.50, 0.88, 0.54], max: 60,  pickup: 8 },
  heavy:   { id: 'heavy',   name: 'Heavy',   css: '#c08cff', color: [0.75, 0.55, 1.0],  pickup: 12, max: 30 },
};

export const SLOTS = {
  kinetic: { id: 'kinetic', name: 'Kinetic', kind: 'weapon', icon: '◆' },
  energy:  { id: 'energy',  name: 'Energy',  kind: 'weapon', icon: '◇' },
  power:   { id: 'power',   name: 'Power',   kind: 'weapon', icon: '✦' },
  helmet:  { id: 'helmet',  name: 'Helmet',  kind: 'armor',  icon: '⌂' },
  arms:    { id: 'arms',    name: 'Gauntlets', kind: 'armor', icon: '⊐' },
  chest:   { id: 'chest',   name: 'Chest',   kind: 'armor',  icon: '⬔' },
  legs:    { id: 'legs',    name: 'Legs',    kind: 'armor',  icon: '⋔' },
  class:   { id: 'class',   name: 'Class Item', kind: 'armor', icon: '❖' },
};
export const SLOT_ORDER = ['kinetic', 'energy', 'power', 'helmet', 'arms', 'chest', 'legs', 'class'];
export const WEAPON_SLOTS = ['kinetic', 'energy', 'power'];
export const ARMOR_SLOTS = ['helmet', 'arms', 'chest', 'legs', 'class'];

/** The six character stats, in display order. */
export const STATS = [
  { id: 'resilience', name: 'Resilience', desc: 'Raises maximum health and reduces flinch.' },
  { id: 'mobility',   name: 'Mobility',   desc: 'Increases movement and strafe speed.' },
  { id: 'recovery',   name: 'Recovery',   desc: 'Shields recharge sooner and faster.' },
  { id: 'discipline', name: 'Discipline', desc: 'Reduces grenade cooldown.' },
  { id: 'intellect',  name: 'Intellect',  desc: 'Reduces Super cooldown.' },
  { id: 'strength',   name: 'Strength',   desc: 'Reduces melee cooldown.' },
];
export const STAT_IDS = STATS.map((s) => s.id);

// ---------------------------------------------------------------- power curve
export const POWER = {
  start: 100,
  softCap: 300,     // world drops stop climbing here
  powerfulCap: 350, // "powerful" rewards climb to here
  pinnacleCap: 370, // pinnacle rewards only
};

/** Stat tier 0..10 from a 0..100 stat total. */
export const statTier = (v) => Math.max(0, Math.min(10, Math.floor(v / 10)));

/**
 * Destiny-style power delta scaling.
 * Being under the recommended power hurts fast; being over helps, but with a ceiling.
 */
export function powerDamageOut(playerPower, activityPower) {
  const d = playerPower - activityPower;
  const m = d >= 0 ? 1 + d * 0.010 : 1 + d * 0.022;
  return Math.max(0.20, Math.min(1.55, m));
}
export function powerDamageIn(playerPower, activityPower) {
  const d = playerPower - activityPower;
  const m = d >= 0 ? 1 - d * 0.007 : 1 - d * 0.020;
  return Math.max(0.55, Math.min(2.6, m));
}

/** XP required to reach the next character level. */
export function xpForLevel(level) {
  return Math.floor(600 + level * 340 + Math.pow(level, 1.75) * 24);
}

/** How much a drop's power can exceed the player's, given the reward tier. */
export function dropPower(playerPower, tier, rng = Math.random) {
  const p = Math.round(playerPower);
  if (tier === 'pinnacle') {
    return Math.min(POWER.pinnacleCap, p + 3 + Math.floor(rng() * 3));
  }
  if (tier === 'powerful') {
    return Math.min(POWER.powerfulCap, p + 2 + Math.floor(rng() * 3));
  }
  // World drops: mostly sidegrades, occasionally a small bump, hard-stopped at the soft cap.
  const roll = -3 + Math.floor(rng() * 6); // -3..+2
  return Math.min(POWER.softCap, Math.max(1, p + roll));
}

/** Element that counters a shield — matching element shreds it. */
export const SHIELD_MATCH_MULT = 3.0;
export const SHIELD_MISMATCH_MULT = 0.45;

export const CRIT_TEXT_COLOR = '#ffe08a';
